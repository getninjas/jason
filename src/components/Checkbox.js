import PropTypes from 'prop-types';
import React, { Component, createRef } from 'react';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  title: PropTypes.string,
  placeholder: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  initialValue: PropTypes.any,
  values: PropTypes.array,
  style: PropTypes.string,
  minLength: PropTypes.number,
  maxLength: PropTypes.number,
};

const defaultProps = {
  placeholder: '',
  required: false,
  title: '',
  type: 'text',
  initialValue: '',
  values: [],
  style: 'form__input',
  minLength: 3,
  maxLength: 255,
};

export default class Checkbox extends Component {
  constructor(props) {
    super(props);

    this.state = {
      values: this.props.values,
    };

    this.ref = createRef();
    this.onChange = this.onChange.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.mask = null;
    this.normalizeInputCheck = this.normalizeInputCheck.bind(this);
  }

  normalizeInputCheck(value) {
    const inputChecked = value.filter(input => input.checked)
      .map((input) => {
        if (input.value === 'OTHER') {
          return input.textOther;
        }

        return input.databaseId;
      });

    return inputChecked;
  }

  onChange(evt) {
    const idNumberEvt = parseInt(evt.target.id, 10);
    const values = this.state.values.map((eachItem) => {
      const itemToSave = eachItem;

      if (eachItem.databaseId === idNumberEvt) {
        itemToSave.checked = evt.target.checked;

        if (eachItem.value === 'OTHER' && evt.target.type !== 'checkbox') {
          itemToSave.textOther = evt.target.value;
        }
      }

      return Object.assign(eachItem, itemToSave);
    });

    this.props.onFieldChange({
      id: this.props.id,
      value: this.normalizeInputCheck(values),
    });

    this.setState({ values });
  }

  onBlur(evt) {
    const data = {
      target: {
        checked: true,
        id: evt.target.getAttribute('data-id'),
        value: evt.target.value,
      },
    };

    this.onChange(data);
  }

  render() {
    const {
      name,
      title,
      style,
      required,
    } = this.props;

    return (
      <ul required={required ? 'true' : 'false'}>
        {this.props.values.map((elem, idx) => (
          <li className='form__check' key={`${elem.databaseId}-${idx}`} htmlFor={elem.databaseId}>
            <input
              type='checkbox'
              id={elem.databaseId}
              name={name}
              className={style}
              defaultChecked={!elem.checked ? false : elem.checked}
              onChange={this.onChange} />

            <label key={`${elem.databaseId}-${idx}`} htmlFor={elem.databaseId}>
              {elem.value === 'OTHER' ? (
                <input type="text" className={style} disabled={!elem.checked} data-id={elem.databaseId} ref={this.ref} onBlur={this.onBlur} />
              ) : elem.value}
            </label>
          </li>
          ))}
      </ul>
    );
  }
}

Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;
