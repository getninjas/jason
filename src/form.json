{
  "form": {
    "title": "Form Title",
    "databaseId": 81,
    "zipcodeUrlService": "http://www.mocky.io/v2/5afaebc92f00000e00f7c654",
    "steps": [
      {
        "title": "General data",
        "breadcrumbTitle": "Your request",
        "buttonText": "Start request",
        "headerMarkup": "<h2 class='widget__title section-title--high-contrast'>Lorem Ipsum is simply dummy text <small>It has survived not</small></h2>",
        "fields": [
          {
            "title": "Qual é o tipo de serviço que você precisa?",
            "placeholder": "Placeholder",
            "required": true,
            "id": "1_id",
            "name": "1_name_select",
            "type": "select",
            "value": 7117,
            "values": [
              {
                "databaseId": 7117,
                "value": "Construção"
              },
              {
                "databaseId": 7118,
                "value": "Reformas"
              },
              {
                "databaseId": 7119,
                "value": "Instalações"
              },
              {
                "databaseId": 7120,
                "value": "Troca"
              },
              {
                "databaseId": 7121,
                "value": "Outros"
              }
            ]
          },
          {
            "title": "Especifique melhor a metragem do serviço",
            "required": true,
            "id": "2_id",
            "name": "2_name_text",
            "type": "text",
            "placeholder": "Bora pra action",
            "value": "AHUAHuhhHUHAU",
            "values": []
          },
          {
            "title": "Informações adicionais",
            "required": false,
            "type": "textarea",
            "id": "3_id",
            "name": "3_name_textarea",
            "placeholder": "Descreva o que você precisa",
            "value": null,
            "values": []
          },
          {
            "required": true,
            "title": "Você precisa de serviços adicionais?",
            "placeholder": "Você precisa de serviços adicionais?",
            "id": "0_id",
            "name": "1_name_select",
            "type": "checkbox",
            "value": null,
            "values": [
              {
                "databaseId": 6792,
                "value": "Cozinhar"
              },
              {
                "databaseId": 7639,
                "value": "OTHER"
              }
            ]
          }
        ]
      },
      {
        "title": "User Data",
        "breadcrumbTitle": "Your data",
        "buttonText": "Finish",
        "headerMarkup": "<div class='form__alert-message space-element-large'><img src='https://www.getninjas.com.br/assets/request/phone_call.svg' class='form__icon' alt='Lorem Ipsum is simply dummy text of the printing and typesetting industry.'><h2 class='form__info'>Almost there<small>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's<strong>standard dummy text ever since the 1500s</strong></small></h2></div>",
        "fields": [
          {
            "title": "Zipcode",
            "required": true,
            "type": "zipcode",
            "id": "zip_code",
            "name": "zipcode",
            "placeholder": "00000-000",
            "value": "04904-060",
            "wrapperClassName": "form__field input",
            "values": []
          },
          {
            "title": "DDD + Celular",
            "id": "phone",
            "name": "phone",
            "type": "phone",
            "placeholder": "(__) _____-____",
            "required": true,
            "value": "(11) 96311-2689",
            "values": []
          }
        ]
      }
    ]
  }
}
