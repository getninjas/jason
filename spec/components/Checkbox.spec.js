import React from 'react';
import renderer from 'react-test-renderer';
import Checkbox from '../../src/components/Checkbox';
import { form } from '../../src/form.json';
import { enzymeConfig, mount } from '../enzymeConfig';

enzymeConfig();

function createNodeMock(element) {
  if (element.type === 'checkbox') {
    return {
      addEventListener() {},
      values: '',
    };
  }

  return null;
}

const checkbox = form.steps[0].fields[2];

describe('Checkbox', () => {
  const commonProps = {
    id: 'idTest',
    name: 'nameTest',
    onFieldChange: () => {},
    onFieldBlur: () => {},
    placeholder: 'placeholderTest',
    type: 'checkbox',
    required: true,
  };

  it('render component', () => {
    const options = { createNodeMock };
    const component = renderer.create(
      <Checkbox
        {...commonProps}
        values={[checkbox.values[0]]}
      />, options,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  // it('changes input checked .onChage event', () => {
  //   const onChange = jest.fn();

  //   const wrapper = mount(
  //     <Checkbox
  //       {...commonProps}
  //       values={[checkbox.values[0]]}
  //       onFieldChange={onChange}
  //     />,
  //   );

  //   expect(wrapper.instance().inputChecked).toEqual({});

  //   wrapper.find('input').simulate('change', { target: { checked: true, id: 6792, getAttribute: () => 'Cozinhar' } });

  //   expect(wrapper.instance().props.onFieldChange).toBeCalled();
  //   expect(wrapper.instance().inputChecked).toEqual({ 6792: 6792 });
  // });

  // it('changes input checked .onBlurCheckbox event', () => {
  //   const onBlur = jest.fn();

  //   const wrapper = mount(
  //     <Checkbox
  //       {...commonProps}
  //       values={[checkbox.values[0]]}
  //       onFieldBlur={onBlur}
  //     />,
  //   );

  //   wrapper.find('input').simulate('blur');

  //   expect(wrapper.instance().props.onFieldBlur).toBeCalled();
  // });

  // describe('with input text', () => {
  //   it('render component', () => {
  //     const options = { createNodeMock };
  //     const component = renderer.create(
  //       <Checkbox
  //         {...commonProps}
  //         values={[checkbox.values[5]]}
  //       />, options,
  //     );

  //     const tree = component.toJSON();

  //     expect(tree).toMatchSnapshot();
  //   });

  //   it('checkbox with OTHER value need to have an input text', () => {
  //     const onBlur = jest.fn();

  //     const wrapper = mount(
  //       <Checkbox
  //         {...commonProps}
  //         values={[checkbox.values[5]]}
  //         onFieldChange={onBlur}
  //       />,
  //     );

  //     expect(wrapper.instance().inputChecked).toEqual({});
  //     expect(wrapper.find('input[type="text"]').get(0).props.disabled).toBeTruthy();

  //     wrapper.find('input[type="checkbox"]').simulate('change', { target: { checked: true, id: 7639, getAttribute: () => 'OTHER' } });

  //     expect(wrapper.instance().props.onFieldChange).toBeCalled();
  //     expect(wrapper.find('input[type="text"]').get(0).props.disabled).toBeFalsy();
  //     expect(wrapper.instance().inputChecked).toEqual({ 7639: '' });
  //   });

  //   it('checkbox with OTHER value need to have an input text', () => {
  //     const wrapper = mount(
  //       <Checkbox
  //       {...commonProps}
  //       values={[checkbox.values[5]]}
  //       />,
  //     );

  //     expect(wrapper.instance().inputChecked).toEqual({});

  //     wrapper.find('input[type="checkbox"]').simulate('change', { target: { checked: true, id: 7639, getAttribute: () => 'OTHER' } });
  //     wrapper.find('input[type="text"]').simulate('blur', { target: { value: 'Teste value', getAttribute: () => 7639 } });

  //     expect(wrapper.instance().inputChecked).toEqual({ 7639: 'Teste value' });
  //   });
  // });
});
