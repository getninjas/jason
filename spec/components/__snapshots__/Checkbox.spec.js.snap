// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render component 1`] = `
<ul
  required="true"
>
  <li
    className="form__check"
    htmlFor={6792}
  >
    <input
      className="form__input"
      data-input-value="Cozinhar"
      id={6792}
      name="nameTest"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor={6792}
    >
      Cozinhar
    </label>
  </li>
</ul>
`;

exports[`Checkbox with input text render component 1`] = `
<ul
  required="true"
>
  <li
    className="form__check"
    htmlFor={7639}
  >
    <input
      className="form__input"
      data-input-value="OTHER"
      id={7639}
      name="nameTest"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      htmlFor={7639}
    >
      <input
        className="form__input"
        data-id={7639}
        disabled={true}
        onBlur={[Function]}
        type="text"
      />
    </label>
  </li>
</ul>
`;
